README.md
# ```sanzo```

# Color Palettes Based on the Works of Sanzo Wada
An R package by JRMA Maasch (2019). <a href="https://CRAN.R-project.org/package=sanzo" target="_blank">View on CRAN.</a>

<img src="https://github.com/jmaasch/sanzo/blob/master/dev/sanzo_dictionary/sanzo_amazon1.jpg" width="410" align="right"/>
<img src="https://github.com/jmaasch/sanzo/blob/master/dev/sanzo_dictionary/sanzo1.jpg" width="410" align="right"/>

### Table of contents

1. **[About](#about):** Origins and inspirations.
2. **[Usage](#usage):** &nbsp; [Install.](#installation) &nbsp; | &nbsp; [Docs.](#view-documentation) &nbsp; | &nbsp; [Demo.](#view-palette-demos-using-sanzodemo-functions) &nbsp; | &nbsp; [Base R.](#use-with-base-R) &nbsp; | &nbsp; [ggplot2.](#use-with-ggplot2) &nbsp; | &nbsp; [Gradient.](#use-as-a-continuous-color-palette) &nbsp; | &nbsp; [Metadata.](#access-palette-metadata-with-sanzoinfo-functions) &nbsp; | &nbsp; [Customize.](#customize)
3. **[ggplot2 gallery](#ggplot2-gallery):** &nbsp; [Scatter plots.](#scatter-plots) &nbsp; | [&nbsp; Bar plots.](#bar-plots) &nbsp; | &nbsp; [Violin plots.](#violin-plots) &nbsp; | &nbsp; [2D Density plots.](#2D-Density-plots) &nbsp;
4. **[View full suite](#view-full-suite):** &nbsp; [Duo palettes.](#duo-palettes) &nbsp; | &nbsp; [Trio palettes.](#trio-palettes) &nbsp; | &nbsp; [Quad palettes.](#quad-palettes)

# About

Inspired by the art and color research of Sanzo Wada, his <a href="http://seigensha.com/en/2016/11/01/978-4-86152-247-5/" target="_blank">Dictionary Of Color Combinations</a>, and the incredible <a href="https://github.com/dblodorn/sanzo-wada" target="_blank">interactive site</a> by Dain M. Blodorn Kim, this package brings Wada's color combinations to R for easy use in data visualizations. This package honors 60 of Wada's color combinations: 20 duos, 20 trios, and 20 quads.

> Sanzo Wada (1883-1967) was an artist, teacher, costume and kimono designer during a turbulent time in avant-garde Japanese art and cinema. Wada was ahead of his time in developing traditional and Western influenced color combinations, helping to lay the foundations for contemporary color research. Based on his original 6-volume work from the 1930s, this book offers 348 color combinations with CMYK and Pantone numbers, as attractive and sensuous as the book’s own design. – *A Dictionary Of Color Combinations*

<p align="center">   
<img src="https://github.com/jmaasch/sanzo/blob/master/dev/sanzo_dictionary/sanzo2.jpg" width="700" align="middle"/>
</p>

This package is licensed under the GNU General Public License v3.0 (<a href="https://choosealicense.com/licenses/gpl-3.0/" target="_blank">GPL-3</a>).

&#8593; [return to top](#sanzo)

# Usage

### Installation.

```R
# Install from CRAN (recommended).
install.packages("sanzo")

# Alternatively, install development version.
# Note: the package devtools must be installed but need not be loaded.
devtools::install_github("jmaasch/sanzo")

# Load package.
library(sanzo)
```

### View documentation.

View this <a href="http://rpubs.com/razel/sanzo_vignette" target="_blank">vignette</a> for additional package documentation, or precede any function by a question mark to access description, arguments, return value, and usage suggestions.

```R
# Access function documentation.
?sanzo.info2()
?sanzo.trio()
```

### View palette demos using ```sanzo.demo``` functions.

Each color combination can be demo'd via base R scatter and bar plots. To view the output of these functions, see 
**[Full suite](#full-suite):** [Duo palettes.](#duo-palettes) | [Trio palettes.](#trio-palettes) | [Quad palettes.](#quad-palettes)

```R

# Print demo plots for duos.
sanzo.demo2()

# Print demo plots for trios.
sanzo.demo3()

# Print demo plots for quads.
sanzo.demo4()

# Print demo plots for all palettes.
sanzo.demo.all()
```

*Optional*: for easier viewing, try gridding up prior to running ```sanzo.demo``` functions. **NOTE:** This will alter ```par``` settings.

```R

# Set up gridded layout, if desired:
graphics::layout(matrix(1:4, nrow = 2))
graphics::par(mar = c(2, 2, 2, 2))

```

### Use with base R.

All examples use data from the ```datasets``` package and should be replicable.

 ```R
# Assign palette to a name if desired.
my_duo <- sanzo.duo("c229")
my_trio <- sanzo.trio("c223")
my_quad <- sanzo.quad("c252")

# Use palette that has been assigned to a name.
plot(iris$Sepal.Length, iris$Sepal.Width, col = my_quad)

# Use directly.
plot(iris$Sepal.Length, iris$Sepal.Width, col = sanzo.duo("c085"))
 ```
 
### Use with ```ggplot2```.

Use with ```scale_fill_manual``` or ```scale_color_manual```. All examples use data from the ```datasets``` package and should be replicable.

```R
ggplot(iris, aes(fill = Species, y = Sepal.Width, x = Sepal.Length)) + 
  geom_bar(stat = "identity") +
  scale_fill_manual(values = sanzo.trio("c223"))
```

### Use as a continuous color palette.

By default, ```sanzo``` palettes are discrete. There are multiple ways to achieve a gradient. For example, using ```scale_fill_gradientn``` or ```scale_color_gradientn``` in ```ggplot2``` allows you to use any ```sanzo``` palette as a continuous gradient. See example visualizations [here.](#gradient)

```R
ggplot(faithfuld, aes(waiting, eruptions)) +
   geom_raster(aes(fill = density), interpolate = TRUE) +
   scale_fill_gradientn(colors = sanzo.duo("c102")) 
```

### Access palette metadata with ```sanzo.info``` functions.

These functions return a data frame containing long-form names, short-form IDs, hexadecimal values, and links to Dain M. Blodorn Kim's https://sanzo-wada.dmbk.io for further information.

```R
duo_info_df <- sanzo.info2()

trio_info_df <- sanzo.info3()

quad_info_df <- sanzo.info4()

info_df <- sanzo.info.all()

# Two ways to print a data frame.
print(sanzo.info2())
knitr::kable(sanzo.info2())
```

Duo, trio, and quad data are stored in exported lists. To quickly access palette names or hexadecimals only:

```R
# Access palette names.
names(duos)
names(trios)
names(quads)

# Print hex values for a given palette to the console.
# Example output = [1] "#ff616b" "#000831"
sanzo.duo("c006")
```

<img src="https://github.com/jmaasch/sanzo/blob/master/dev/ggplot2demos/concat.png" width="500" align="right"/>

### Customize.

Concatenate multiple palettes to make a custom palette or cherry-pick individual colors. All hexadecimals are provided by the ```sanzo.demo``` functions and ```sanzo.info``` functions.

 ```R
c033 <- sanzo.duo("c033")
c095 <- sanzo.duo("c095")
custom_quad <- c(c033, c095)
```

For a more muted color palette, select an alpha less than 1. In the plot at right, ```alpha = 5/6```.

&#8593; [return to top](#sanzo)

## ```ggplot2``` gallery

Many of the plots in this gallery additionally use the ```ggthemes``` and ```cowplot``` packages.

### Scatter plots.

|   |   |
|:-------------------------:|:-------------------------:|
|**Combination 205, ```alpha = 3/4.```**|**Combination 194, ```alpha = 3/4.```**|
|![scatter1](https://github.com/jmaasch/sanzo/blob/master/dev/ggplot2demos/scatter1.png)|![scatter2](https://github.com/jmaasch/sanzo/blob/master/dev/ggplot2demos/scatter2.png)|
|**Combination 121, ```alpha = 3/4.```**|**Combination 172, ```alpha = 3/4.```**|
|![scatter3](https://github.com/jmaasch/sanzo/blob/master/dev/ggplot2demos/scatter3.png)|![scatter4](https://github.com/jmaasch/sanzo/blob/master/dev/ggplot2demos/scatter5.png)|

### Violin plots.

<p align="center">   
<img src="https://github.com/jmaasch/sanzo/blob/master/dev/ggplot2demos/violin.png" width="750" align="middle"/>
</p>

### Bar plots.

|   |   |
|:-------------------------:|:-------------------------:|
|![bar1](https://github.com/jmaasch/sanzo/blob/master/dev/ggplot2demos/bar_ggplot2.png)|![bar2](https://github.com/jmaasch/sanzo/blob/master/dev/ggplot2demos/bar_stacked_ggplot2.png)|

### 2D Density plots.

<img src="https://github.com/jmaasch/sanzo/blob/master/dev/ggplot2demos/gradient.png" align="middle"/>

See how to use ```sanzo``` color combinations as continuous palettes [here.](#use-as-a-continuous-color-palette)

&#8593; [return to top](#sanzo)

# View full suite

### Duo palettes.

|   |   |
|:-------------------------:|:-------------------------:|
|![demo1](https://github.com/jmaasch/sanzo/blob/master/dev/demos2/Unknown-1.png)|![demo2](https://github.com/jmaasch/sanzo/blob/master/dev/demos2/Unknown-2.png)| 
|![demo3](https://github.com/jmaasch/sanzo/blob/master/dev/demos2/Unknown-3.png)|![demo4](https://github.com/jmaasch/sanzo/blob/master/dev/demos2/Unknown-4.png)|
|![demo5](https://github.com/jmaasch/sanzo/blob/master/dev/demos2/Unknown-5.png)|![demo6](https://github.com/jmaasch/sanzo/blob/master/dev/demos2/Unknown-6.png)|
|![demo7](https://github.com/jmaasch/sanzo/blob/master/dev/demos2/Unknown-7.png)|![demo8](https://github.com/jmaasch/sanzo/blob/master/dev/demos2/Unknown-8.png)|
|![demo9](https://github.com/jmaasch/sanzo/blob/master/dev/demos2/Unknown-9.png)|![demo10](https://github.com/jmaasch/sanzo/blob/master/dev/demos2/Unknown-10.png)|

&#8593; [return to top](#sanzo)

### Trio palettes.

|   |   |
|:-------------------------:|:-------------------------:|
|![demo1](https://github.com/jmaasch/sanzo/blob/master/dev/demos3/Unknown-1.png)|![demo2](https://github.com/jmaasch/sanzo/blob/master/dev/demos3/Unknown-2.png)| 
|![demo3](https://github.com/jmaasch/sanzo/blob/master/dev/demos3/Unknown-3.png)|![demo4](https://github.com/jmaasch/sanzo/blob/master/dev/demos3/Unknown-4.png)|
|![demo5](https://github.com/jmaasch/sanzo/blob/master/dev/demos3/Unknown-5.png)|![demo6](https://github.com/jmaasch/sanzo/blob/master/dev/demos3/Unknown-6.png)|
|![demo7](https://github.com/jmaasch/sanzo/blob/master/dev/demos3/Unknown-7.png)|![demo8](https://github.com/jmaasch/sanzo/blob/master/dev/demos3/Unknown-8.png)|
|![demo9](https://github.com/jmaasch/sanzo/blob/master/dev/demos3/Unknown-9.png)|![demo10](https://github.com/jmaasch/sanzo/blob/master/dev/demos3/Unknown-10.png)|

&#8593; [return to top](#sanzo)

### Quad palettes.

|   |   |
|:-------------------------:|:-------------------------:|
|![demo1](https://github.com/jmaasch/sanzo/blob/master/dev/demos4/Unknown-1.png)|![demo2](https://github.com/jmaasch/sanzo/blob/master/dev/demos4/Unknown-2.png)| 
|![demo3](https://github.com/jmaasch/sanzo/blob/master/dev/demos4/Unknown-3.png)|![demo4](https://github.com/jmaasch/sanzo/blob/master/dev/demos4/Unknown-4.png)|
|![demo5](https://github.com/jmaasch/sanzo/blob/master/dev/demos4/Unknown-5.png)|![demo6](https://github.com/jmaasch/sanzo/blob/master/dev/demos4/Unknown-6.png)|
|![demo7](https://github.com/jmaasch/sanzo/blob/master/dev/demos4/Unknown-7.png)|![demo8](https://github.com/jmaasch/sanzo/blob/master/dev/demos4/Unknown-8.png)|
|![demo9](https://github.com/jmaasch/sanzo/blob/master/dev/demos4/Unknown-9.png)|![demo10](https://github.com/jmaasch/sanzo/blob/master/dev/demos4/Unknown-10.png)|

&#8593; [return to top](#sanzo)


NEWS.md
# NEWS

## ```sanzo``` 0.1.0
* First release submitted to CRAN.
* Submitted 7 January 2020.


R/duos.R
#' List of duos
#'
#' List containing all 20 duo palettes, defined by
#' hexadecimal values.
#'
#' @export
#'
duos <- list(
  c006 = c("#ff616b", "#000831"),
  c007 = c("#ff5200", "#b3e8c2"),
  c008 = c("#c2612c", "#b8b8ff"),
  c033 = c("#b85e00", "#1b3644"),
  c035 = c("#b08699", "#a10b2b"),
  c047 = c("#c9303e", "#bfabcc"),
  c070 = c("#b85e00", "#405416"),
  c071 = c("#a90636", "#d99e73"),
  c074 = c("#b5ffc2", "#2dbc94"),
  c079 = c("#651300", "#2dbc94"),
  c085 = c("#c74300", "#008aa1"),
  c091 = c("#5c2c45", "#c05200"),
  c095 = c("#681916", "#06004f"),
  c102 = c("#ebd999", "#c05200"),
  c103 = c("#c2612c", "#2d0060"),
  c106 = c("#008aa1", "#06004f"),
  c114 = c("#ffab00", "#008aa1"),
  c118 = c("#e0b81f", "#362304"),
  c119 = c("#a6e6db", "#0d2b52"),
  c229 = c("#172713", "#b5d1cc")
)

#' Generate two-colored palettes
#'
#' Generate two-colored palettes.
#'
#' @param palette_name The short ID for the palette, e.g. "c006". For full list of duo IDs, use sanzo.info2().
#'
#' @return A vector of hexademicals of length 2.
#'
#' @examples
#' # Assign palette to a name.
#' my_palette <- sanzo.duo("c229")
#'
#' # Concatenate two duos for a custom quad.
#' c033 <- sanzo.duo("c033")
#' c095 <- sanzo.duo("c095")
#' custom_quad <- c(c033, c095)
#'
#' # Use with base R.
#' plot(iris$Sepal.Width,
#'      iris$Sepal.Length,
#'      col = sanzo.duo("c085"))
#'
#' # For examples of use with ggplot2, see https://github.com/jmaasch/sanzo.
#'
#' # For examples of use as a gradient, see https://github.com/jmaasch/sanzo.
#'
#' @export
#'
sanzo.duo <- function(palette_name) {
  palette <- duos[[palette_name]]
  return(palette)
}


R/quads.R
#' List of quads
#'
#' List containing all 20 quad palettes, defined by
#' hexadecimal values.
#'
#' @export
quads <- list(
  c249 = c("#681916", "#c0b490", "#d1bd19", "#417777"),
  c252 = c("#e62e73", "#b85e00", "#baa600", "#2dbc94"),
  c263 = c("#a93400", "#f2ad78", "#b5ffc2", "#1b3644"),
  c268 = c("#b08699", "#b85e00", "#172713", "#bfffe6"),
  c271 = c("#b90078", "#94ff94", "#2dbc94", "#0f261f"),
  c273 = c("#ffb3f0", "#6f0043", "#9b5348", "#b5d1cc"),
  c279 = c("#b85e00", "#f59994", "#c0b490", "#0d2b52"),
  c283 = c("#a7374b", "#730f1f", "#85b857", "#6bffb3"),
  c289 = c("#f2ff26", "#bdf226", "#202d85", "#06004f"),
  c293 = c("#b85e00", "#b5ffc2", "#65a98f", "#40c945"),
  c296 = c("#f5f5b8", "#d99e73", "#5e4017", "#1b3644"),
  c302 = c("#ffb852", "#c0b490", "#bfffe6", "#008aa1"),
  c310 = c("#f5f5b8", "#f2ad78", "#718600", "#172713"),
  c321 = c("#b08699", "#f5f5b8", "#172713", "#96bfe6"),
  c329 = c("#ffb852", "#bfabcc", "#340059", "#1b3644"),
  c330 = c("#bcd382", "#bfffe6", "#96bfe6", "#2dbc94"),
  c332 = c("#ff7399", "#d50c42", "#172713", "#00592e"),
  c339 = c("#d99e73", "#de4500", "#a6e6db", "#202d85"),
  c341 = c("#ff616b", "#328e13", "#172713", "#a6e6db"),
  c348 = c("#bcd382", "#328e13", "#172713", "#340059")
)

#' Generate four-colored palettes
#'
#' Generate four-colored palettes.
#'
#' @param palette_name The short ID for the palette, e.g. "c263". For full list of quad IDs, use sanzo.info4().
#'
#' @return A vector of hexademicals of length 3.
#'
#' @examples
#' # Assign palette to a name.
#' my_palette <- sanzo.quad("c252")
#'
#' # Concatenate two quads for a custom eight-colored palette.
#' c348 <- sanzo.quad("c348")
#' c341 <- sanzo.quad("c341")
#' custom_eight <- c(c348, c341)
#'
#' # Use with base R.
#' plot(iris$Sepal.Width,
#'      iris$Sepal.Length,
#'      col = sanzo.quad("c341"))
#'
#' # For examples of use with ggplot2, see https://github.com/jmaasch/sanzo.
#'
#' # For examples of use as a gradient, see https://github.com/jmaasch/sanzo.
#'
#' @export
#'
sanzo.quad <- function(palette_name) {
  palette <- quads[[palette_name]]
  return(palette)
}


R/trios.R
#' List of trios
#'
#' List containing all 20 trio palettes, defined by
#' hexadecimal values.
#'
#' @export
#'
trios <- list(
  c121 = c("#6c2b11", "#d99e73", "#405416"),
  c139 = c("#96bfe6", "#000831", "#b5d1cc"),
  c142 = c("#9e194d", "#baa600", "#96bfe6"),
  c148 = c("#d1bd19", "#ffab00", "#29bdad"),
  c149 = c("#d1bd19", "#ff5200", "#0f261f"),
  c157 = c("#6f0043", "#d1bd19", "#4f8fe6"),
  c163 = c("#ffe600", "#b5ffc2", "#008aa1"),
  c171 = c("#730f1f", "#ff8c00", "#b3e8c2"),
  c172 = c("#c2612c", "#008aa1", "#3400a3"),
  c174 = c("#ffa6d9", "#bfabcc", "#6b2e63"),
  c194 = c("#fa2b00", "#ffcfc4", "#4f8fe6"),
  c204 = c("#b319ab", "#c2612c", "#a6e6db"),
  c205 = c("#730f1f", "#f59994", "#2619d1"),
  c207 = c("#9b5348", "#b3e8c2", "#000000"),
  c223 = c("#b08699", "#d99e73", "#b5ffc2"),
  c226 = c("#5c2c45", "#ffb852", "#2619d1"),
  c232 = c("#d60036", "#f2ad78", "#000831"),
  c239 = c("#b08699", "#c4bf33", "#b3e8c2"),
  c240 = c("#ff788c", "#ffff00", "#29bdad"),
  c343 = c("#a93400", "#ebd999", "#505423")
)

#' Generate three-colored palettes
#'
#' Generate three-colored palettes.
#'
#' @param palette_name The short ID for the palette, e.g. "c121". For full list of trio IDs, use sanzo.info3().
#'
#' @return A vector of hexademicals of length 3.
#'
#' @examples
#' # Assign palette to a name.
#' my_palette <- sanzo.trio("c223")
#'
#' # Concatenate two trios for a custom six-colored palette.
#' c207 <- sanzo.trio("c207")
#' c226 <- sanzo.trio("c226")
#' custom_six <- c(c207, c226)
#'
#' # Use with base R.
#' plot(iris$Sepal.Width,
#'      iris$Sepal.Length,
#'      col = sanzo.trio("c343"))
#'
#' # For examples of use with ggplot2, see https://github.com/jmaasch/sanzo.
#'
#' # For examples of use as a gradient, see https://github.com/jmaasch/sanzo.
#'
#' @export
#'
sanzo.trio <- function(palette_name) {
  palette <- trios[[palette_name]]
  return(palette)
}


R/info.R
#' See metadata for all duo palettes
#'
#' Create data frame containing long-form names, short-form IDs,
#' hexadecimal values, and links to Dain M. Blodorn Kim's
#' https://sanzo-wada.dmbk.io for all duo palettes.
#'
#' @return Return data frame containing long-form names, short-form IDs,
#' hex values, and URLs for all duos.
#'
#' @examples
#' duo_info_df <- sanzo.info2()
#' print(sanzo.info2())
#'
#' @export
#'
sanzo.info2 <- function() {

  long_names <- c("Combination 6", "Combination 7", "Combination 8", "Combination 33",
                  "Combination 35", "Combination 47", "Combination 70", "Combination 71",
                  "Combination 74", "Combination 79", "Combination 85", "Combination 91",
                  "Combination 95", "Combination 102", "Combination 103", "Combination 106",
                  "Combination 114", "Combination 118", "Combination 119", "Combination 229")
  short_names <- c("c006", "c007", "c008", "c033", "c035", "c047", "c070", "c071", "c074", "c079",
                   "c085", "c091", "c095", "c102", "c103", "c106", "c114", "c118", "c119", "c229")
  hex <- c("#ff616b, #000831", "#ff5200, #b3e8c2", "#c2612c, #b8b8ff", "#b85e00, #1b3644",
           "#b08699, #a10b2b", "#c9303e, #bfabcc", "#b85e00, #405416", "#a90636, #d99e73",
           "#b5ffc2, #2dbc94", "#651300, #2dbc94", "#c74300, #008aa1", "#5c2c45, #c05200",
           "#681916, #06004f", "#ebd999, #c05200", "#c2612c, #2d0060", "#008aa1, #06004f",
           "#ffab00, #008aa1", "#e0b81f, #362304", "#a6e6db, #0d2b52", "#172713, #b5d1cc")
  url <- c("https://sanzo-wada.dmbk.io/combination/6", "https://sanzo-wada.dmbk.io/combination/7",
           "https://sanzo-wada.dmbk.io/combination/8", "https://sanzo-wada.dmbk.io/combination/33",
           "https://sanzo-wada.dmbk.io/combination/35", "https://sanzo-wada.dmbk.io/combination/47",
           "https://sanzo-wada.dmbk.io/combination/70", "https://sanzo-wada.dmbk.io/combination/71",
           "https://sanzo-wada.dmbk.io/combination/74", "https://sanzo-wada.dmbk.io/combination/79",
           "https://sanzo-wada.dmbk.io/combination/85", "https://sanzo-wada.dmbk.io/combination/91",
           "https://sanzo-wada.dmbk.io/combination/95", "https://sanzo-wada.dmbk.io/combination/102",
           "https://sanzo-wada.dmbk.io/combination/103", "https://sanzo-wada.dmbk.io/combination/106",
           "https://sanzo-wada.dmbk.io/combination/114", "https://sanzo-wada.dmbk.io/combination/118",
           "https://sanzo-wada.dmbk.io/combination/119", "https://sanzo-wada.dmbk.io/combination/229")

  df <- data.frame(long_names, short_names, hex, url)
  colnames(df) <- c("DUO PALETTE", "ID", "HEX VALUES", "URL")
  return(df)

}

#' See metadata for all trio palettes
#'
#' Create data frame containing long-form names, short-form IDs,
#' hexadecimal values, and links to Dain M. Blodorn Kim's
#' https://sanzo-wada.dmbk.io for all trio palettes.
#'
#' @return Return data frame containing long-form names, short-form IDs,
#' hex values, and URLs for all trios.
#'
#' @examples
#' trio_info_df <- sanzo.info3()
#' print(sanzo.info3())
#'
#' @export
#'
sanzo.info3 <- function() {

  long_names <- c("Combination 121", "Combination 139", "Combination 142", "Combination 148",
                  "Combination 149", "Combination 157", "Combination 163", "Combination 171",
                  "Combination 172", "Combination 174", "Combination 194", "Combination 204",
                  "Combination 205", "Combination 207", "Combination 223", "Combination 226",
                  "Combination 232", "Combination 239", "Combination 240", "Combination 343")
  short_names <- c("c121", "c139", "c142", "c148", "c149", "c157", "c163", "c171", "c172", "c174",
                   "c194", "c204", "c205", "c207", "c223", "c226", "c232", "c239", "c240", "c343")
  hex <- c("#6c2b11, #d99e73, #405416", "#96bfe6, #000831, #b5d1cc", "#9e194d, #baa600, #96bfe6",
           "#d1bd19, #ffab00, #29bdad", "#d1bd19, #ff5200, #0f261f", "#6f0043, #d1bd19, #4f8fe6",
           "#ffe600, #b5ffc2, #008aa1", "#730f1f, #ff8c00, #b3e8c2", "#c2612c, #008aa1, #3400a3",
           "#ffa6d9, #bfabcc, #6b2e63", "#fa2b00, #ffcfc4, #4f8fe6", "#b319ab, #c2612c, #a6e6db",
           "#730f1f, #f59994, #2619d1", "#9b5348, #b3e8c2, #000000", "#b08699, #d99e73, #b5ffc2",
           "#5c2c45, #ffb852, #2619d1", "#d60036, #f2ad78, #000831", "#b08699, #c4bf33, #b3e8c2",
           "#ff788c, #ffff00, #29bdad", "#a93400, #ebd999, #505423")
  url <- c("https://sanzo-wada.dmbk.io/combination/121", "https://sanzo-wada.dmbk.io/combination/139",
           "https://sanzo-wada.dmbk.io/combination/142", "https://sanzo-wada.dmbk.io/combination/148",
           "https://sanzo-wada.dmbk.io/combination/149", "https://sanzo-wada.dmbk.io/combination/157",
           "https://sanzo-wada.dmbk.io/combination/163", "https://sanzo-wada.dmbk.io/combination/171",
           "https://sanzo-wada.dmbk.io/combination/172", "https://sanzo-wada.dmbk.io/combination/174",
           "https://sanzo-wada.dmbk.io/combination/194", "https://sanzo-wada.dmbk.io/combination/204",
           "https://sanzo-wada.dmbk.io/combination/205", "https://sanzo-wada.dmbk.io/combination/207",
           "https://sanzo-wada.dmbk.io/combination/223", "https://sanzo-wada.dmbk.io/combination/226",
           "https://sanzo-wada.dmbk.io/combination/232", "https://sanzo-wada.dmbk.io/combination/239",
           "https://sanzo-wada.dmbk.io/combination/240", "https://sanzo-wada.dmbk.io/combination/343")

  df <- data.frame(long_names, short_names, hex, url)
  colnames(df) <- c("TRIO PALETTE", "ID", "HEX VALUES", "URL")
  return(df)

}

#' See metadata for all quad palettes
#'
#' Create data frame containing long-form names, short-form IDs,
#' hexadecimal values, and links to Dain M. Blodorn Kim's
#' https://sanzo-wada.dmbk.io for all quad palettes.
#'
#' @return Return data frame containing long-form names, short-form IDs,
#' hex values, and URLs for all quads
#'
#' @examples
#' quad_info_df <- sanzo.info4()
#' print(sanzo.info4())
#'
#' @export
#'
sanzo.info4 <- function() {

  long_names <- c("Combination 249", "Combination 252", "Combination 263", "Combination 268",
                  "Combination 271", "Combination 273", "Combination 279", "Combination 283",
                  "Combination 289", "Combination 293", "Combination 296", "Combination 302",
                  "Combination 310", "Combination 321", "Combination 329", "Combination 330",
                  "Combination 332", "Combination 339", "Combination 341", "Combination 348")
  short_names <- c("c249", "c252", "c263", "c268", "c271", "c273", "c279", "c283", "c289", "c293",
                   "c296", "c302", "c310", "c321", "c329", "c330", "c332", "c339", "c341", "c348")
  hex <- c("#681916, #c0b490, #d1bd19, #417777", "#e62e73, #b85e00, #baa600, #2dbc94",
           "#a93400, #f2ad78, #b5ffc2, #1b3644", "#b08699, #b85e00, #172713, #bfffe6",
           "#b90078, #94ff94, #2dbc94, #0f261f", "#ffb3f0, #6f0043, #9b5348, #b5d1cc",
           "#b85e00, #f59994, #c0b490, #0d2b52", "#a7374b, #730f1f, #85b857, #6bffb3",
           "#f2ff26, #bdf226, #202d85, #06004f", "#b85e00, #b5ffc2, #65a98f, #40c945",
           "#f5f5b8, #d99e73, #5e4017, #1b3644", "#ffb852, #c0b490, #bfffe6, #008aa1",
           "#f5f5b8, #f2ad78, #718600, #172713", "#b08699, #f5f5b8, #172713, #96bfe6",
           "#ffb852, #bfabcc, #340059, #1b3644", "#bcd382, #bfffe6, #96bfe6, #2dbc94",
           "#ff7399, #d50c42, #172713, #00592e", "#d99e73, #de4500, #a6e6db, #202d85",
           "#ff616b, #328e13, #172713, #a6e6db", "#bcd382, #328e13, #172713, #340059")
  url <- c("https://sanzo-wada.dmbk.io/combination/249", "https://sanzo-wada.dmbk.io/combination/252",
           "https://sanzo-wada.dmbk.io/combination/263", "https://sanzo-wada.dmbk.io/combination/268",
           "https://sanzo-wada.dmbk.io/combination/271", "https://sanzo-wada.dmbk.io/combination/273",
           "https://sanzo-wada.dmbk.io/combination/279", "https://sanzo-wada.dmbk.io/combination/283",
           "https://sanzo-wada.dmbk.io/combination/289", "https://sanzo-wada.dmbk.io/combination/293",
           "https://sanzo-wada.dmbk.io/combination/296", "https://sanzo-wada.dmbk.io/combination/302",
           "https://sanzo-wada.dmbk.io/combination/310", "https://sanzo-wada.dmbk.io/combination/321",
           "https://sanzo-wada.dmbk.io/combination/329", "https://sanzo-wada.dmbk.io/combination/330",
           "https://sanzo-wada.dmbk.io/combination/332", "https://sanzo-wada.dmbk.io/combination/339",
           "https://sanzo-wada.dmbk.io/combination/341", "https://sanzo-wada.dmbk.io/combination/348")

  df <- data.frame(long_names, short_names, hex, url)
  colnames(df) <- c("QUAD PALETTE", "ID", "HEX VALUES", "URL")
  return(df)

}

#' See metadata for all palettes
#'
#' Create data frame containing long-form names, short-form IDs,
#' hexadecimal values, and links to Dain M. Blodorn Kim's
#' https://sanzo-wada.dmbk.io for all sanzo palettes.
#'
#' @return Return data frame containing long-form names, short-form IDs,
#' hex values, and URLs for all sanzo palettes.
#'
#' @examples
#' info_df <- sanzo.info.all()
#' print(sanzo.info.all())
#'
#' @export
#'
sanzo.info.all <- function() {

  d <- sanzo.info2()
  t <- sanzo.info3()
  q <- sanzo.info4()

  df <- merge(d, t, all = TRUE)
  df <- merge(df, q, all = TRUE)
  return(df)

}
